File: 02200/BOJ_2273.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    private static final int INF = 257;
    private void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int[][] arr = new int[n][n];
        for (int i = 0; i < n; i++) Arrays.fill(arr[i], INF);
        while (m-->0) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken())-1;
            int b = Integer.parseInt(st.nextToken())-1;
            arr[a][b] = 1;
        }
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (i == j || arr[i][k] == INF || arr[k][j] == INF) continue;
                    arr[i][j] = Math.min(arr[i][j], arr[i][k]+arr[k][j]);
                }
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            int outCnt = 0;
            int inCnt = 0;
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                if (arr[i][j] != INF && arr[j][i] != INF) {
                    System.out.println(-1);
                    return;
                }
                if (arr[i][j] != INF) outCnt++;
                if (arr[j][i] != INF) inCnt++;
            }
            sb.append(inCnt+1).append(' ').append(n-outCnt).append('\n');
        }
        System.out.print(sb);
    }
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: 02200/BOJ_2273.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer
import kotlin.math.min

const val INF = 257
fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    var st = StringTokenizer(readLine())
    val n = st.nextToken().toInt()
    var m = st.nextToken().toInt()
    val arr = Array(n){IntArray(n){INF}}
    repeat(m) {
        st = StringTokenizer(readLine())
        val a = st.nextToken().toInt()-1
        val b = st.nextToken().toInt()-1
        arr[a][b] = 1
    }
    for (k in 0 until n) {
        for (i in 0 until n) {
            for (j in 0 until n) {
                if (i == j || arr[i][k] == INF || arr[k][j] == INF) continue
                arr[i][j] = min(arr[i][j], arr[i][k] + arr[k][j])
            }
        }
    }
    val sb = StringBuilder()
    for (i in 0 until n) {
        var outCnt = 0
        var inCnt = 0
        for (j in 0 until n) {
            if (i == j) continue
            if (arr[i][j] != INF && arr[j][i] != INF) return println(-1)
            if (arr[i][j] != INF) outCnt++
            if (arr[j][i] != INF) inCnt++
        }
        sb.append(inCnt+1).append(' ').append(n-outCnt).append('\n')
    }
    print(sb)
}
